File: src/factories/TypeFactory.ts
import ts from "typescript";

import { TypePredicator } from "../utils/TypePredicator";

import { Resolved } from "../Resolved";

export namespace TypeFactory {
  export const isFunction = (type: ts.Type): boolean =>
    getFunction(type) !== null;

  const getFunction = (type: ts.Type) => {
    const node = type.symbol?.declarations?.[0];
    if (node === undefined) return null;

    return ts.isFunctionLike(node)
      ? node
      : ts.isPropertyAssignment(node) || ts.isPropertyDeclaration(node)
        ? ts.isFunctionLike(node.initializer)
          ? node.initializer
          : null
        : null;
  };

  export const getReturnType =
    (checker: ts.TypeChecker) =>
    (type: ts.Type) =>
    (name: string): ts.Type | null => {
      // FIND TO-JSON METHOD
      const symbol: ts.Symbol | undefined = checker.getPropertyOfType(
        type,
        name,
      );
      if (!symbol) return null;
      else if (!symbol.valueDeclaration) return null;

      // GET FUNCTION DECLARATION
      const functor: ts.Type = checker.getTypeOfSymbolAtLocation(
        symbol,
        symbol.valueDeclaration,
      );

      // RETURNS THE RETURN-TYPE
      const signature: ts.Signature | undefined = checker.getSignaturesOfType(
        functor,
        ts.SignatureKind.Call,
      )[0];
      return signature ? signature.getReturnType() : null;
    };

  export const getFullName =
    (checker: ts.TypeChecker) =>
    (type: ts.Type, symbol?: ts.Symbol): string => {
      // PRIMITIVE
      symbol ??= type.aliasSymbol ?? type.symbol;
      if (symbol === undefined) return checker.typeToString(type);

      // UNION OR INTERSECT
      if (
        type.aliasSymbol === undefined &&
        TypePredicator.isUnionOrIntersection(type)
      ) {
        const joiner: string = TypePredicator.isIntersection(type)
          ? " & "
          : " | ";
        return type.types
          .map((child) => getFullName(checker)(child))
          .join(joiner);
      }

      //----
      // SPECIALIZATION
      //----
      const name: string = get_name(symbol);

      // CHECK GENERIC
      const generic: readonly ts.Type[] = type.aliasSymbol
        ? type.aliasTypeArguments || []
        : checker.getTypeArguments(type as ts.TypeReference);
      return generic.length
        ? name === "Promise"
          ? getFullName(checker)(generic[0]!)
          : `${name}<${generic
              .map((child) => getFullName(checker)(child))
              .join(", ")}>`
        : name;
    };

  const explore_name =
    (decl: ts.Node) =>
    (name: string): string =>
      ts.isModuleBlock(decl)
        ? explore_name(decl.parent.parent)(
            `${decl.parent.name.getFullText().trim()}.${name}`,
          )
        : name;

  const get_name = (symbol: ts.Symbol): string => {
    const parent = symbol.getDeclarations()?.[0]?.parent;
    return parent
      ? explore_name(parent)(symbol.escapedName.toString())
      : "__type";
  };

  export const keyword = (
    type:
      | "void"
      | "any"
      | "unknown"
      | "boolean"
      | "number"
      | "bigint"
      | "string",
  ) => {
    return ts.factory.createKeywordTypeNode(
      type === "void"
        ? ts.SyntaxKind.VoidKeyword
        : type === "any"
          ? ts.SyntaxKind.AnyKeyword
          : type === "unknown"
            ? ts.SyntaxKind.UnknownKeyword
            : type === "boolean"
              ? ts.SyntaxKind.BooleanKeyword
              : type === "number"
                ? ts.SyntaxKind.NumberKeyword
                : type === "bigint"
                  ? ts.SyntaxKind.BigIntKeyword
                  : ts.SyntaxKind.StringKeyword,
    );
  };

  export const clone = (type: ts.Type): Resolved<ts.Type> =>
    cloneType(type, 0, new Set());
}

const cloneType = (x: any, depth: number, visited: Set<object>): any => {
  if (x === undefined || depth > 4) return undefined;
  else if (typeof x === "object")
    if (x === null) return null;
    else if (Array.isArray(x))
      return x.map((y) => cloneType(y, depth + 1, visited));
    else {
      visited.add(x);
      const entries = (() => {
        try {
          return Object.entries(x);
        } catch {
          return undefined;
        }
      })();
      if (entries === undefined) return undefined;
      return Object.fromEntries(
        entries
          .filter(
            ([k, y]) =>
              k !== "parent" &&
              typeof y !== "function" &&
              !(typeof y === "object" && y !== null && visited.has(y) === true),
          )
          .map(([k, y]) => {
            // if (typeof y === "object" && y !== null) console.log(k, get_uid(y));
            return [k, cloneType(y, depth + 1, visited)];
          }),
      );
    }
  return x;
};

// function get_uid(obj: object | null | undefined): number {
//   // NO UID EXISTS, THEN ISSUE ONE.
//   if (obj instanceof Object) {
//     if (obj.hasOwnProperty("__get_m_iUID") === false) {
//       const uid: number = ++__s_iUID;
//       Object.defineProperty(obj, "__get_m_iUID", {
//         value: function (): number {
//           return uid;
//         },
//       });
//     }

//     // RETURNS
//     return (obj as IObject).__get_m_iUID();
//   } else if (obj === undefined) return -1;
//   // is null
//   else return 0;
// }

// interface IObject {
//   readonly __get_m_iUID: () => number;
// }
// let __s_iUID: number = 0;

File: packages/typia/native/core/factories/TypeFactory.go
package factories

import (
	"strings"

	shimast "github.com/microsoft/typescript-go/shim/ast"
	shimchecker "github.com/microsoft/typescript-go/shim/checker"
)

type typeFactoryNamespace struct{}

var TypeFactory = typeFactoryNamespace{}

type TypeFactory_GetReturnTypeOfClassMethodProps struct {
	Checker  *shimchecker.Checker
	Class    *shimchecker.Type
	Function string
}

type TypeFactory_GetFullNameProps struct {
	Checker            *shimchecker.Checker
	Type               *shimchecker.Type
	Symbol             *shimast.Symbol
	AliasTypeArguments *bool
}

var typeFactory_factory = shimast.NewNodeFactory(shimast.NodeFactoryHooks{})

func (typeFactoryNamespace) IsFunction(t *shimchecker.Type) bool {
	return TypeFactory.GetFunction(t) != nil
}

func (typeFactoryNamespace) GetFunction(t *shimchecker.Type) *shimast.Node {
	if t == nil {
		return nil
	}
	symbol := t.Symbol()
	if symbol == nil || len(symbol.Declarations) == 0 {
		return nil
	}
	node := symbol.Declarations[0]
	if shimast.IsFunctionLike(node) {
		return node
	}
	if shimast.IsPropertyAssignment(node) {
		initializer := node.AsPropertyAssignment().Initializer
		if shimast.IsFunctionLike(initializer) {
			return initializer
		}
		return nil
	}
	if shimast.IsPropertyDeclaration(node) {
		initializer := node.AsPropertyDeclaration().Initializer
		if shimast.IsFunctionLike(initializer) {
			return initializer
		}
	}
	return nil
}

func (typeFactoryNamespace) GetReturnTypeOfClassMethod(props TypeFactory_GetReturnTypeOfClassMethodProps) *shimchecker.Type {
	if props.Checker == nil || props.Class == nil {
		return nil
	}
	symbol := props.Checker.GetPropertyOfType(props.Class, props.Function)
	if symbol == nil || symbol.ValueDeclaration == nil {
		return nil
	}
	functor := props.Checker.GetTypeOfSymbolAtLocation(symbol, symbol.ValueDeclaration)
	signatures := props.Checker.GetSignaturesOfType(functor, shimchecker.SignatureKindCall)
	if len(signatures) == 0 {
		return nil
	}
	return props.Checker.GetReturnTypeOfSignature(signatures[0])
}

func (typeFactoryNamespace) GetFullName(props TypeFactory_GetFullNameProps) string {
	if props.Checker == nil || props.Type == nil {
		return ""
	}
	symbol := props.Symbol
	if symbol == nil {
		symbol = props.Type.Symbol()
	}
	if symbol == nil {
		if props.Type.IsUnion() || props.Type.IsIntersection() {
			joiner := " | "
			if props.Type.IsIntersection() {
				joiner = " & "
			}
			children := props.Type.Types()
			names := make([]string, 0, len(children))
			for _, child := range children {
				names = append(names, TypeFactory.GetFullName(TypeFactory_GetFullNameProps{
					Checker: props.Checker,
					Type:    child,
				}))
			}
			return strings.Join(names, joiner)
		}
		return props.Checker.TypeToString(props.Type)
	}
	name := typeFactory_get_name(symbol)
	generic := props.Checker.GetTypeArguments(props.Type)
	if len(generic) == 0 {
		return name
	}
	if name == "Promise" {
		return TypeFactory.GetFullName(TypeFactory_GetFullNameProps{
			Checker: props.Checker,
			Type:    generic[0],
		})
	}
	names := make([]string, 0, len(generic))
	for _, child := range generic {
		names = append(names, TypeFactory.GetFullName(TypeFactory_GetFullNameProps{
			Checker: props.Checker,
			Type:    child,
		}))
	}
	return name + "<" + strings.Join(names, ", ") + ">"
}

func typeFactory_explore_name(node *shimast.Node, name string) string {
	if node != nil && shimast.IsModuleBlock(node) && node.Parent != nil && node.Parent.Parent != nil {
		parentName := ""
		if node.Parent.Name() != nil {
			parentName = strings.TrimSpace(node.Parent.Name().Text())
		}
		if parentName != "" {
			return typeFactory_explore_name(node.Parent.Parent, parentName+"."+name)
		}
	}
	return name
}

func typeFactory_get_name(symbol *shimast.Symbol) string {
	if symbol == nil || len(symbol.Declarations) == 0 || symbol.Declarations[0].Parent == nil {
		return "__type"
	}
	return typeFactory_explore_name(symbol.Declarations[0].Parent, symbol.Name)
}

func (typeFactoryNamespace) Keyword(t string) *shimast.Node {
	var kind shimast.KeywordTypeSyntaxKind
	switch t {
	case "void":
		kind = shimast.KindVoidKeyword
	case "any":
		kind = shimast.KindAnyKeyword
	case "unknown":
		kind = shimast.KindUnknownKeyword
	case "boolean":
		kind = shimast.KindBooleanKeyword
	case "number":
		kind = shimast.KindNumberKeyword
	case "bigint":
		kind = shimast.KindBigIntKeyword
	default:
		kind = shimast.KindStringKeyword
	}
	return typeFactory_factory.NewKeywordTypeNode(kind)
}
